File: vrpc-py/src/vuer_vrpc/zdata.py
"""
ZData - Extensible type encoding system for binary serialization.

Provides encoding/decoding for numpy arrays and allows registration of custom types.
Optional extensions for PyTorch tensors, PIL images, and safetensors are available.

Basic usage:
    >>> import numpy as np
    >>> from vmp_py import ZData
    >>> arr = np.array([1, 2, 3])
    >>> encoded = ZData.encode(arr)
    >>> decoded = ZData.decode(encoded)

With optional extensions:
    >>> from vmp_py.extensions import torch_support
    >>> import torch
    >>> tensor = torch.randn(3, 3)
    >>> encoded = ZData.encode(tensor)

Custom types:
    >>> class Point:
    ...     def __init__(self, x, y): self.x, self.y = x, y
    >>> def encode_point(p):
    ...     return {"ztype": "custom.Point", "b": f"{p.x},{p.y}".encode()}
    >>> def decode_point(z):
    ...     x, y = map(float, z["b"].decode().split(","))
    ...     return Point(x, y)
    >>> ZData.register_type("custom.Point", encode_point, decode_point, Point)
"""

from typing import Any, Callable, Optional
from .type_registry import TYPE_REGISTRY, ZDataDict, TypeEncoder, TypeDecoder

# Import built-in types (numpy is always available)
from . import builtin_types  # noqa: F401


class ZData:
    """
    Main interface for ZData encoding/decoding.

    Uses the global TYPE_REGISTRY which can be extended by:
    1. Importing optional extensions (torch_support, image_support, etc.)
    2. Registering custom types with ZData.register_type()
    3. Directly accessing TYPE_REGISTRY in external libraries

    Examples:
        >>> import numpy as np
        >>> arr = np.array([1, 2, 3])
        >>> encoded = ZData.encode(arr)
        >>> decoded = ZData.decode(encoded)

        >>> # Register custom type
        >>> class Point:
        ...     def __init__(self, x, y): self.x, self.y = x, y
        >>> ZData.register_type(
        ...     "custom.Point",
        ...     lambda p: {"ztype": "custom.Point", "b": f"{p.x},{p.y}".encode()},
        ...     lambda z: Point(*map(float, z["b"].decode().split(","))),
        ...     type_class=Point
        ... )
    """

    @staticmethod
    def encode(data: Any) -> Any:
        """
        Encode data to ZData format if a matching encoder is registered.

        Args:
            data: Data to encode (numpy array, torch tensor, PIL image, etc.)

        Returns:
            ZDataDict if data type is registered, otherwise returns data unchanged

        Example:
            >>> import numpy as np
            >>> arr = np.array([1, 2, 3])
            >>> encoded = ZData.encode(arr)
            >>> encoded['ztype']
            'numpy.ndarray'
        """
        return TYPE_REGISTRY.encode(data)

    @staticmethod
    def decode(zdata: Any) -> Any:
        """
        Decode ZData back to original type.

        Args:
            zdata: ZData dictionary or plain data

        Returns:
            Decoded object if zdata is valid ZData, otherwise returns zdata unchanged

        Raises:
            TypeError: If zdata has an unknown ztype

        Example:
            >>> encoded = ZData.encode(np.array([1, 2, 3]))
            >>> decoded = ZData.decode(encoded)
            >>> type(decoded)
            <class 'numpy.ndarray'>
        """
        return TYPE_REGISTRY.decode(zdata)

    @staticmethod
    def is_zdata(data: Any) -> bool:
        """
        Check if data is a ZData encoded object.

        Args:
            data: Data to check

        Returns:
            True if data is a ZData dictionary

        Example:
            >>> arr = np.array([1, 2, 3])
            >>> encoded = ZData.encode(arr)
            >>> ZData.is_zdata(encoded)
            True
            >>> ZData.is_zdata(arr)
            False
        """
        return TYPE_REGISTRY.is_zdata(data)

    @staticmethod
    def get_ztype(data: Any) -> Optional[str]:
        """
        Get the ztype of a ZData object.

        Args:
            data: Data to check

        Returns:
            The ztype string if data is ZData, None otherwise

        Example:
            >>> encoded = ZData.encode(np.array([1, 2, 3]))
            >>> ZData.get_ztype(encoded)
            'numpy.ndarray'
        """
        return TYPE_REGISTRY.get_ztype(data)

    @staticmethod
    def register_type(
        type_name: str,
        encoder: TypeEncoder,
        decoder: TypeDecoder,
        type_class: Optional[type] = None,
        type_checker: Optional[Callable[[Any], bool]] = None
    ):
        """
        Register a custom type for encoding/decoding.

        Args:
            type_name: Unique identifier for this type (e.g., "custom.MyType")
            encoder: Function that encodes data to ZDataDict
            decoder: Function that decodes ZDataDict back to original type
            type_class: Optional type class for direct type checking
            type_checker: Optional function for custom type checking

        Example:
            >>> class Point:
            ...     def __init__(self, x, y): self.x, self.y = x, y
            ...
            >>> def encode_point(p):
            ...     return {"ztype": "custom.Point", "b": f"{p.x},{p.y}".encode()}
            ...
            >>> def decode_point(z):
            ...     x, y = map(float, z["b"].decode().split(","))
            ...     return Point(x, y)
            ...
            >>> ZData.register_type("custom.Point", encode_point, decode_point, Point)
            >>> point = Point(1.0, 2.0)
            >>> encoded = ZData.encode(point)
            >>> decoded = ZData.decode(encoded)
        """
        TYPE_REGISTRY.register(type_name, encoder, decoder, type_class, type_checker)

    @staticmethod
    def list_types() -> list[str]:
        """
        List all registered type names.

        Returns:
            List of registered ztype names

        Example:
            >>> ZData.list_types()
            ['numpy.ndarray', 'torch.Tensor', ...]
        """
        return TYPE_REGISTRY.list_registered_types()

File: vrpc-ts/src/zdata.ts
/**
 * ZData - Extensible type encoding system for binary serialization.
 *
 * Provides encoding/decoding for TypedArrays and allows registration of custom types.
 *
 * Basic usage:
 * ```typescript
 * import { ZData } from '@vuer-ai/vmp-ts';
 *
 * const data = new Float32Array([1, 2, 3]);
 * const encoded = ZData.encode(data);
 * const decoded = ZData.decode(encoded);
 * ```
 *
 * Custom types:
 * ```typescript
 * class Point {
 *   constructor(public x: number, public y: number) {}
 * }
 *
 * ZData.register_type(
 *   "custom.Point",
 *   (p) => ({ ztype: "custom.Point", b: new TextEncoder().encode(`${p.x},${p.y}`) }),
 *   (z) => {
 *     const [x, y] = new TextDecoder().decode(z.b).split(",").map(Number);
 *     return new Point(x, y);
 *   },
 *   Point
 * );
 * ```
 */

import type { ZDataDict, TypeEncoder, TypeDecoder, TypeChecker } from './types.js';
import { TYPE_REGISTRY } from './type_registry.js';

// Import built-in types (TypedArrays are always available)
import './builtin_types.js';

/**
 * Main interface for ZData encoding/decoding.
 *
 * Uses the global TYPE_REGISTRY which can be extended by:
 * 1. Registering custom types with ZData.register_type()
 * 2. Directly accessing TYPE_REGISTRY in external libraries
 */
export class ZData {
  /**
   * Encode data to ZData format if a matching encoder is registered.
   *
   * @param data - Data to encode (TypedArray, custom objects, etc.)
   * @returns ZDataDict if data type is registered, otherwise returns data unchanged
   *
   * @example
   * ```typescript
   * const arr = new Float32Array([1, 2, 3]);
   * const encoded = ZData.encode(arr);
   * console.log(encoded.ztype); // "Float32Array"
   * ```
   */
  static encode(data: unknown): unknown {
    return TYPE_REGISTRY.encode(data);
  }

  /**
   * Decode ZData back to original type.
   *
   * @param zdata - ZData dictionary or plain data
   * @returns Decoded object if zdata is valid ZData, otherwise returns zdata unchanged
   * @throws TypeError if zdata has an unknown ztype
   *
   * @example
   * ```typescript
   * const encoded = ZData.encode(new Float32Array([1, 2, 3]));
   * const decoded = ZData.decode(encoded);
   * console.log(decoded); // Float32Array [1, 2, 3]
   * ```
   */
  static decode(zdata: unknown): unknown {
    return TYPE_REGISTRY.decode(zdata);
  }

  /**
   * Check if data is a ZData encoded object.
   *
   * @param data - Data to check
   * @returns True if data is a ZData dictionary
   *
   * @example
   * ```typescript
   * const arr = new Float32Array([1, 2, 3]);
   * const encoded = ZData.encode(arr);
   * console.log(ZData.is_zdata(encoded)); // true
   * console.log(ZData.is_zdata(arr)); // false
   * ```
   */
  static is_zdata(data: unknown): data is ZDataDict {
    return TYPE_REGISTRY.isZData(data);
  }

  /**
   * Get the ztype of a ZData object.
   *
   * @param data - Data to check
   * @returns The ztype string if data is ZData, undefined otherwise
   *
   * @example
   * ```typescript
   * const encoded = ZData.encode(new Float32Array([1, 2, 3]));
   * console.log(ZData.get_ztype(encoded)); // "Float32Array"
   * ```
   */
  static get_ztype(data: unknown): string | undefined {
    return TYPE_REGISTRY.getZType(data);
  }

  /**
   * Register a custom type for encoding/decoding.
   *
   * @param typeName - Unique identifier for this type (e.g., "custom.MyType")
   * @param encoder - Function that encodes data to ZDataDict
   * @param decoder - Function that decodes ZDataDict back to original type
   * @param typeClass - Optional type class/constructor for direct type checking
   * @param typeChecker - Optional function for custom type checking
   *
   * @example
   * ```typescript
   * class Point {
   *   constructor(public x: number, public y: number) {}
   * }
   *
   * ZData.register_type(
   *   "custom.Point",
   *   (p: Point) => ({
   *     ztype: "custom.Point",
   *     b: new TextEncoder().encode(`${p.x},${p.y}`)
   *   }),
   *   (z: ZDataDict) => {
   *     const [x, y] = new TextDecoder().decode(z.b as Uint8Array)
   *       .split(",").map(Number);
   *     return new Point(x, y);
   *   },
   *   Point
   * );
   *
   * const point = new Point(1.0, 2.0);
   * const encoded = ZData.encode(point);
   * const decoded = ZData.decode(encoded);
   * ```
   */
  static register_type(
    typeName: string,
    encoder: TypeEncoder,
    decoder: TypeDecoder,
    typeClass?: Function,
    typeChecker?: TypeChecker
  ): void {
    TYPE_REGISTRY.register(typeName, encoder, decoder, typeClass, typeChecker);
  }

  /**
   * List all registered type names.
   *
   * @returns List of registered ztype names
   *
   * @example
   * ```typescript
   * const types = ZData.list_types();
   * console.log(types); // ['Float32Array', 'Float64Array', ...]
   * ```
   */
  static list_types(): string[] {
    return TYPE_REGISTRY.listRegisteredTypes();
  }

  /**
   * Identity function that wraps data with a ztype tag without encoding.
   *
   * Useful for adding type information to data without performing binary encoding.
   * The data is kept as-is, only wrapped in a ZDataDict with the specified ztype.
   *
   * @param ztype - Type identifier to tag the data with
   * @param data - Data to wrap (kept as-is)
   * @returns ZDataDict with ztype and original data
   *
   * @example
   * ```typescript
   * // Tag data without encoding
   * const tagged = ZData.identity('custom.MyType', { x: 1, y: 2 });
   * // Returns: { ztype: 'custom.MyType', data: { x: 1, y: 2 } }
   *
   * // Useful for polymorphic data structures
   * const geometries = [
   *   ZData.identity('sphere', { radius: 1 }),
   *   ZData.identity('box', { size: [1, 2, 3] }),
   *   ZData.identity('cylinder', { radius: 0.5, height: 2 }),
   * ];
   * ```
   */
  static identity(ztype: string, data: unknown): ZDataDict {
    return {
      ztype,
      data,
    };
  }
}

/**
 * Export the TYPE_REGISTRY for advanced use cases
 */
export { TYPE_REGISTRY } from './type_registry.js';
